File: public/api/face_detect_fast.php
<?php
require __DIR__ . '/_util.php';
$cfg = cfg();

// 1) Lấy ảnh đầu vào: ưu tiên file upload 'image'; nếu không có thì chụp từ ESP32
$tmp = tempnam(sys_get_temp_dir(), 'det_') . '.jpg';
if (!empty($_FILES['image']['tmp_name'])) {
    move_uploaded_file($_FILES['image']['tmp_name'], $tmp);
} else {
    $ip = get_esp_ip();
    $ok = false;
    foreach ($cfg['snapshot_paths'] as $p) {
        $url = "http://{$ip}{$p}";
        $bytes = http_get_bytes($url, 6);
        if ($bytes && str_starts_with($bytes, "\xFF\xD8")) {
            file_put_contents($tmp, $bytes);
            $ok = true;
            break;
        }
    }
    if (!$ok) bad('Không chụp được ảnh từ ESP32');
}

// 2) Gọi Python xử lý DETECT ONLY (nhanh)
$cmd = escapeshellcmd($cfg['python_bin']) . ' ' .
    escapeshellarg($cfg['tools_dir'] . '/face_detect_only.py') . ' ' .
    '--image ' . escapeshellarg($tmp);
exec($cmd . ' 2>&1', $out, $code);
@unlink($tmp);

if ($code !== 0) {
    bad("Python error ($code): " . implode("\n", $out), 500);
}

// 3) Python in ra JSON -> pass through
header('Content-Type: application/json; charset=utf-8');
header('Cache-Control: no-store');
echo implode("\n", $out);

File: public/api/face_detect_fast.py
"""
Face Detect Fast API - Python Version
Chuyển đổi từ face_detect_fast.php
"""
import os
import tempfile
import subprocess
from flask import request
from _util import cfg, get_esp_ip, json_out, bad, http_get_bytes

def face_detect_fast():
    """Phát hiện khuôn mặt nhanh (detect only)"""
    config = cfg()
    
    # 1) Lấy ảnh đầu vào
    tmp_file = tempfile.NamedTemporaryFile(suffix='.jpg', delete=False)
    tmp_path = tmp_file.name
    tmp_file.close()
    
    try:
        if 'image' in request.files:
            file = request.files['image']
            file.save(tmp_path)
        else:
            ip = get_esp_ip(request.args)
            ok = False
            for path in config['snapshot_paths']:
                url = f"http://{ip}{path}"
                img_bytes = http_get_bytes(url, 6)
                if img_bytes and img_bytes[:2] == b'\xFF\xD8':
                    with open(tmp_path, 'wb') as f:
                        f.write(img_bytes)
                    ok = True
                    break
            if not ok:
                return bad('Không chụp được ảnh từ ESP32')
        
        # 2) Gọi Python detect only (nhanh)
        cmd = [
            config['python_bin'],
            os.path.join(config['tools_dir'], 'face_detect_only.py'),
            '--image', tmp_path
        ]
        
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=config['python_timeout'])
        
        if result.returncode != 0:
            return bad(f"Python error ({result.returncode}): {result.stderr}", 500)
        
        # 3) Python output JSON
        from flask import Response
        return Response(result.stdout, mimetype='application/json', headers={'Cache-Control': 'no-store'})
        
    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)
